src/main/java/com/minecolonies/structures/helpers/Structure.java

```diff
@@ -84,61 +97,340 @@
+     * <p>
+     * Look into the following director (in order):
+     * - scan
+     * - cache
+     * - schematics folder
+     * - jar
+     * It should be the exact oppsite that the way used to buikd the list.
+     *
+     * @param structureName name of the structure to load
+     * @return the input stream or null
+     */
+    public static InputStream getStream(final String structureName)
+    {
+        final Structures.StructureName sn = new Structures.StructureName(structureName);
+        InputStream inputstream;
+        if (Structures.SCHEMATICS_CACHE.equals(sn.getPrefix()))
+        {
+            return Structure.getStreamFromFolder(Structure.getCachedSchematicsFolder(), structureName);
+        }
+        else if (Structures.SCHEMATICS_SCAN.equals(sn.getPrefix()))
+        {
+            return Structure.getStreamFromFolder(Structure.getClientSchematicsFolder(), structureName);
+        }
+        else if (!Structures.SCHEMATICS_PREFIX.equals(sn.getPrefix()))
+        {
+            return null;
+        }
+        else
+        {
+            //Look in the folder first
+            inputstream = Structure.getStreamFromFolder(Structure.getSchematicsFolder(), structureName);
+            if (inputstream == null && !Configurations.ignoreSchematicsFromJar)
+            {
+                inputstream = Structure.getStreamFromJar(structureName);
+            }
+        }
+
+        if (inputstream == null)
+        {
+            Log.getLogger().warn("Structure: Couldn't find any structure with this name " + structureName);
+        }
+
+        return inputstream;
+    }
+
+    /**
+     * get a input stream for a schematic within a specif folder.
+     *
+     * @param folder        where to load it from.
+     * @param structureName name of the structure to load.
+     * @return the input stream or null
+     */
+    private static InputStream getStreamFromFolder(final File folder, final String structureName)
+    {
+        final File nbtFile = new File(folder.getPath() + "/" + structureName + ".nbt");
+        try
+        {
+            if (folder.exists())
+            {
+                //We need to check that we stay within the correct folder
+                if (!nbtFile.toURI().normalize().getPath().startsWith(folder.toURI().normalize().getPath()))
                 {
-                    inputstream = new FileInputStream(decorationFolder.getPath() + "/" + structureName + ".nbt");
+                    Log.getLogger().error("Structure: Illegal structure name \"" + structureName + "\"");
+                    return null;
                 }
-                else
+                if (nbtFile.exists())
                 {
-                    throw new FileNotFoundException("Unable to find structure: " + structureName);
+                    return new FileInputStream(nbtFile);
                 }
             }
-            catch (final FileNotFoundException e)
+        }
+        catch (final FileNotFoundException e)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)